lib/ClangImporter/ImportDecl.cpp

```diff
@@ -2444,19 +2444,120 @@ namespace {
+          // case we need to bail here.
+          auto cachedResult =
+              Impl.ImportedDecls.find({decl->getCanonicalDecl(), getVersion()});
+          if (cachedResult != Impl.ImportedDecls.end())
+            return cachedResult->second;
+          dc = cast<ExtensionDecl>(parent)
+                   ->getExtendedType()
+                   ->getEnumOrBoundGenericEnum();
+        } else {
+          // Otherwise, this is a non-namesapce decl, so it's owned by its
+          // module in both C++ and Swift meaning we can proceed normally.
+          dc = extensionDC;
+        }
+      }
+      auto *enumDecl = Impl.createDeclWithClangNode<EnumDecl>(
           decl, AccessLevel::Public, loc,
           importedName.getDeclName().getBaseIdentifier(),
           Impl.importSourceLoc(decl->getLocation()), None, nullptr, dc);
-      enumDecl->setMemberLoader(&Impl, 0);
+      if (isa<clang::NamespaceDecl>(decl->getParent()))
+        cast<EnumDecl>(dc)->addMember(enumDecl);
+
+      // We are creating an extension, so put it at the top level. This is done
+      // after creating the enum, though, because we may need the "correctly"
+      // nested decl context above when creating the enum.
+      while (extensionDC->getParent() &&
+             extensionDC->getContextKind() != DeclContextKind::FileUnit)
+        extensionDC = extensionDC->getParent();
+
+      auto *extension = ExtensionDecl::create(Impl.SwiftContext, loc, nullptr,
+                                              {}, extensionDC, nullptr, decl);
+      Impl.SwiftContext.evaluator.cacheOutput(ExtendedTypeRequest{extension},
+                                              enumDecl->getDeclaredType());
+      Impl.SwiftContext.evaluator.cacheOutput(ExtendedNominalRequest{extension},
+                                              std::move(enumDecl));
+      // Keep track of what members we've already added so we don't add the same
+      // member twice. Note: we can't use "ImportedDecls" for this because we
+      // might import a decl that we don't add (for example, if it was a
+      // parameter to another decl).
+      SmallPtrSet<Decl *, 16> addedMembers;
+      for (auto redecl : decl->redecls()) {
+        // This will be re-set as the enum after we return from
+        // "VisitNamespaceDecl."
+        Impl.ImportedDecls[{redecl->getCanonicalDecl(), getVersion()}] =
+            extension;
+
+        // Insert these backwards into "namespaceDecls" so we can pop them off
+        // the end without loosing order.
+        SmallVector<clang::Decl *, 16> namespaceDecls;
+        auto addDeclsBackwards = [&](auto begin, auto end) {
+          int currentSize = namespaceDecls.size();
+          int declCount = currentSize + std::distance(begin, end);
+          namespaceDecls.resize(declCount);
+          for (int index = declCount - 1; index >= currentSize; --index)
+            namespaceDecls[index] = *(begin++);
+        };
```

Comment:
I'd suggest to lean onto `llvm-project/llvm/include/llvm/ADT/STLExtras.h` more.

```
auto addDeclsReversed = [&](auto decls) {
  llvm::append_range(namespaceDecls, llvm::reverse(decls));
};
...
addDeclsReversed(redecl->decls());
...
addDeclsReversed(classTemplate->specializations());
```

Reply:
I think we'll need to wait for the rebranch to use `llvm::append_range`, unfortunately. But I'll add a comment and make the other changes.

Reply:
Also, `DeclContext::decl_iterator` is a forward iterator, and `llvm::reverse` only works on bidirectional iterators. I think the complicated for-loop may, unfortunately, be necessary.